File: apps/api/app/services/criteria/presets.py
from __future__ import annotations
from dataclasses import dataclass
from .registry import CORE_IDS

@dataclass(frozen=True)
class Preset:
    category: str
    weights: dict
    age_overlay_lambda: float

_FLOOR = 0.02  # every core criterion keeps a small non-zero weight

_RAW: dict[str, dict] = {
    "ai_tool": {"trust":.14,"differentiation":.13,"proof_requirement":.12,"value_clarity":.10,
                "perceived_roi":.10,"workflow_fit":.09,"pricing_acceptance":.08,"need_intensity":.08,
                "activation_likelihood":.07,"retention_potential":.05,"shareability":.04},
    "b2b_saas": {"perceived_roi":.15,"workflow_fit":.14,"trust":.12,"switching_willingness":.11,
                 "pricing_acceptance":.10,"proof_requirement":.10,"solution_fit":.09,"differentiation":.06,
                 "activation_likelihood":.05,"retention_potential":.04},
    "consumer_app": {"ease_of_understanding":.15,"activation_likelihood":.14,"shareability":.12,
                     "retention_potential":.11,"value_clarity":.10,"need_intensity":.09,"solution_fit":.08,
                     "repeat_usage_potential":.07,"differentiation":.05},
    "ecommerce_product": {"perceived_roi":.13,"pricing_acceptance":.13,"trust":.12,"value_clarity":.11,
                          "differentiation":.10,"need_intensity":.09,"solution_fit":.08,"proof_requirement":.08,
                          "shareability":.05},
    "education_product": {"trust":.14,"proof_requirement":.12,"perceived_roi":.12,"pricing_acceptance":.11,
                          "need_intensity":.10,"repeat_usage_potential":.09,"solution_fit":.08,"value_clarity":.07},
    "marketplace": {"trust":.14,"need_intensity":.12,"activation_likelihood":.11,"value_clarity":.10,
                    "pricing_acceptance":.09,"differentiation":.09,"retention_potential":.08,"shareability":.07},
    "social_product": {"shareability":.16,"activation_likelihood":.13,"retention_potential":.12,
                       "need_intensity":.10,"ease_of_understanding":.10,"value_clarity":.08,"differentiation":.07},
    "hardware_product": {"perceived_roi":.13,"trust":.12,"proof_requirement":.11,"differentiation":.11,
                         "value_clarity":.10,"pricing_acceptance":.10,"need_intensity":.09,"solution_fit":.08},
    "luxury_product": {"differentiation":.16,"trust":.13,"perceived_roi":.11,"shareability":.11,
                       "need_intensity":.10,"value_clarity":.09,"pricing_acceptance":.06,"retention_potential":.06},
    "generic": {},  # all-floor => uniform
}
_LAMBDA = {"ai_tool":.12,"b2b_saas":.07,"consumer_app":.20,"ecommerce_product":.15,"education_product":.25,
           "marketplace":.15,"social_product":.22,"hardware_product":.15,"luxury_product":.20,"generic":.15}
CATEGORY_IDS = tuple(_RAW.keys())

def resolve_preset(category: str) -> Preset:
    cat = category if category in _RAW else "generic"
    raw = {cid: _RAW[cat].get(cid, _FLOOR) for cid in CORE_IDS}
    total = sum(raw.values())
    weights = {cid: w / total for cid, w in raw.items()}
    return Preset(category=cat, weights=weights, age_overlay_lambda=_LAMBDA[cat])

File: apps/web/lib/server/engine/criteria/presets.ts
/**
 * Category criteria-weight presets — port of
 * apps/api/app/services/criteria/presets.py.
 */

import { CORE_IDS } from "./registry";

export interface Preset {
  category: string;
  weights: Record<string, number>;
  ageOverlayLambda: number;
}

const FLOOR = 0.02; // every core criterion keeps a small non-zero weight

const RAW: Record<string, Record<string, number>> = {
  ai_tool: { trust: 0.14, differentiation: 0.13, proof_requirement: 0.12, value_clarity: 0.1, perceived_roi: 0.1, workflow_fit: 0.09, pricing_acceptance: 0.08, need_intensity: 0.08, activation_likelihood: 0.07, retention_potential: 0.05, shareability: 0.04 },
  b2b_saas: { perceived_roi: 0.15, workflow_fit: 0.14, trust: 0.12, switching_willingness: 0.11, pricing_acceptance: 0.1, proof_requirement: 0.1, solution_fit: 0.09, differentiation: 0.06, activation_likelihood: 0.05, retention_potential: 0.04 },
  consumer_app: { ease_of_understanding: 0.15, activation_likelihood: 0.14, shareability: 0.12, retention_potential: 0.11, value_clarity: 0.1, need_intensity: 0.09, solution_fit: 0.08, repeat_usage_potential: 0.07, differentiation: 0.05 },
  ecommerce_product: { perceived_roi: 0.13, pricing_acceptance: 0.13, trust: 0.12, value_clarity: 0.11, differentiation: 0.1, need_intensity: 0.09, solution_fit: 0.08, proof_requirement: 0.08, shareability: 0.05 },
  education_product: { trust: 0.14, proof_requirement: 0.12, perceived_roi: 0.12, pricing_acceptance: 0.11, need_intensity: 0.1, repeat_usage_potential: 0.09, solution_fit: 0.08, value_clarity: 0.07 },
  marketplace: { trust: 0.14, need_intensity: 0.12, activation_likelihood: 0.11, value_clarity: 0.1, pricing_acceptance: 0.09, differentiation: 0.09, retention_potential: 0.08, shareability: 0.07 },
  social_product: { shareability: 0.16, activation_likelihood: 0.13, retention_potential: 0.12, need_intensity: 0.1, ease_of_understanding: 0.1, value_clarity: 0.08, differentiation: 0.07 },
  hardware_product: { perceived_roi: 0.13, trust: 0.12, proof_requirement: 0.11, differentiation: 0.11, value_clarity: 0.1, pricing_acceptance: 0.1, need_intensity: 0.09, solution_fit: 0.08 },
  luxury_product: { differentiation: 0.16, trust: 0.13, perceived_roi: 0.11, shareability: 0.11, need_intensity: 0.1, value_clarity: 0.09, pricing_acceptance: 0.06, retention_potential: 0.06 },
  generic: {}, // all-floor => uniform
};

const LAMBDA: Record<string, number> = {
  ai_tool: 0.12, b2b_saas: 0.07, consumer_app: 0.2, ecommerce_product: 0.15, education_product: 0.25,
  marketplace: 0.15, social_product: 0.22, hardware_product: 0.15, luxury_product: 0.2, generic: 0.15,
};

export const CATEGORY_IDS: readonly string[] = Object.keys(RAW);

export function resolvePreset(category: string): Preset {
  const cat = category in RAW ? category : "generic";
  const raw: Record<string, number> = {};
  for (const cid of CORE_IDS) raw[cid] = RAW[cat][cid] ?? FLOOR;
  const total = Object.values(raw).reduce((s, w) => s + w, 0);
  const weights: Record<string, number> = {};
  for (const cid of CORE_IDS) weights[cid] = raw[cid] / total;
  return { category: cat, weights, ageOverlayLambda: LAMBDA[cat] };
}
